src/test/java/com/HubertKarw/medical_clinic/Service/PatientJpaServiceTest.java

```diff
@@ -0,0 +1,102 @@
+    PatientJpaRepository patientJpaRepository;
+    PatientJpaService patientJpaService;
+
+    @BeforeEach
+    void setup() {
+        this.patientJpaRepository = Mockito.mock(PatientJpaRepository.class);
+        this.patientJpaService = new PatientJpaService(patientJpaRepository);
+    }
+
+    @Test
+    void getPatient_patientsExist_patientsReturned() {
+        //given
+        User user = new User(1l, "123", "321");
+        Patient patient1 = new Patient(123L, "x1", user, "123", "xyz", "xzy", "123", LocalDate.of(1999, 11, 11));
+        Patient patient2 = new Patient(124L, "x2", user, "124", "xyz", "xzy", "123", LocalDate.of(1999, 12, 11));
+        Patient patient3 = new Patient(125L, "x3", user, "125", "xyz", "xzy", "123", LocalDate.of(1999, 10, 11));
+        PageImpl<Patient> page = new PageImpl<>(List.of(patient1, patient2, patient3));
+        when(patientJpaRepository.findAll(any(Pageable.class))).thenReturn(page);
+        //when
+        List<Patient> result = patientJpaService.getPatients(PageRequest.of(0, 3));
+        //then
+        assertAll(
+                () -> assertEquals(3, result.size()),
+                () -> assertEquals(123, result.get(0).getId()),
+                () -> assertEquals(124, result.get(1).getId()),
+                () -> assertEquals(125, result.get(2).getId())
+        );
+    }
+
+    @Test
+    void getPatient_patientExists_patientReturned() {
+        //given
+        User user = new User(1l, "123", "321");
+        Patient patient = new Patient(123L, "x1", user, "123", "xyz", "xzy", "123", LocalDate.of(1999, 11, 11));
+        when(patientJpaRepository.findByEmail("x1")).thenReturn(Optional.of(patient));
+        //when
+        Patient result = patientJpaService.getPatient("x1");
+        //then
+        assertEquals(patient.getId(), result.getId());
+    }
+
+    @Test
+    void addPatient_validPatient_patientAdded() {
+        //given
+        User user = new User(1l, "123", "321");
+        Patient patient = new Patient(null, "x1", user, "123", "xyz", "xzy", "123", LocalDate.of(1999, 11, 11));
+        Patient patientSaved = new Patient(123L, "x1", user, "123", "xyz", "xzy", "123", LocalDate.of(1999, 11, 11));
+        when(patientJpaRepository.save(any())).thenReturn(patientSaved);
+        //when
+        Patient result = patientJpaService.addPatient(patient);
+        //then
+        assertEquals(patient.getEmail(), patientSaved.getEmail());
+    }
+
+    @Test
+    void removePatient_patientExists_patientRemoved() {
+        //given
+        User user = new User(1l, "123", "321");
+        Patient patient = new Patient(null, "x1", user, "123", "xyz", "xzy", "123", LocalDate.of(1999, 11, 11));
+        Patient patientSaved = new Patient(123L, "x1", user, "123", "xyz", "xzy", "123", LocalDate.of(1999, 11, 11));
+        when(patientJpaRepository.findByEmail(any())).thenReturn(Optional.of(patient));
+        //when
+        patientJpaService.removePatient(patient.getEmail());
```

Comment:
Ta metoda jest voidem wiec nie zwraca nam zadnego rezultatu. W takim przypadku to co mozemy zweryfikowac to czy odpowiednie mocki zostały wywołane -> weryfikacje tego jaki mock powinien zostac wywolany z jaim argumentem mozemy sprawdzic przy pomocy metody verify z Mockito.

Powinnismy sie wiec zastanowic co to znaczy w kontekscie naszego serwisu ze zostal pacjent usuniety -> w kontekscie serwisu usuniecie pacjenta oznacza ze zostala wywolana metoda delete z patientRepository.

Reply:
verify(patientRepository).delete(patient)